[SPARK-14945][PYTHON] SparkSession Python API

andrewor14 opened the pull request:
## What changes were proposed in this pull request?

```
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /__ / .__/\_,_/_/ /_/\_\   version 2.0.0-SNAPSHOT
      /_/

Using Python version 2.7.5 (default, Mar  9 2014 22:15:05)
SparkSession available as 'spark'.
>>> spark
<pyspark.sql.session.SparkSession object at 0x101f3bfd0>
>>> spark.sql("SHOW TABLES").show()
...
+---------+-----------+
|tableName|isTemporary|
+---------+-----------+
|      src|      false|
+---------+-----------+

>>> spark.range(1, 10, 2).show()
+---+
| id|
+---+
|  1|
|  3|
|  5|
|  7|
|  9|
+---+
```

**Note**: This API is NOT complete in its current state. In particular, for now I left out the `conf` and `catalog` APIs, which were added later in Scala. These will be added later before 2.0.
## How was this patch tested?

Python tests.

Commit:
Implement SparkSession + use it in pyspark shell

SparkQA commented:
**[Test build #57191 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57191/consoleFull)** for PR 12746 at commit [`8370767`](https://github.com/apache/spark/commit/8370767759b5191098c87271cda60c6dcbe9fc72).

SparkQA commented:
**[Test build #57191 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57191/consoleFull)** for PR 12746 at commit [`8370767`](https://github.com/apache/spark/commit/8370767759b5191098c87271cda60c6dcbe9fc72).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class SparkSession(object):`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57191/
Test FAILed.

Commit:
Fix style + don't mention 'sc' in shell

SparkQA commented:
**[Test build #57192 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57192/consoleFull)** for PR 12746 at commit [`d6abed1`](https://github.com/apache/spark/commit/d6abed1ba9ec956f0cd3bb063622f783cf1f9825).

SparkQA commented:
**[Test build #57192 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57192/consoleFull)** for PR 12746 at commit [`d6abed1`](https://github.com/apache/spark/commit/d6abed1ba9ec956f0cd3bb063622f783cf1f9825).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57192/
Test FAILed.

Commit:
Fix some of the tests

Commit:
Fix JavaMember has no attribute wrapped

SparkQA commented:
**[Test build #57224 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57224/consoleFull)** for PR 12746 at commit [`2c0ea42`](https://github.com/apache/spark/commit/2c0ea4261d3c1963e76d7de57b19d1f8e18b97f0).

SparkQA commented:
**[Test build #57224 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57224/consoleFull)** for PR 12746 at commit [`2c0ea42`](https://github.com/apache/spark/commit/2c0ea4261d3c1963e76d7de57b19d1f8e18b97f0).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57224/
Test FAILed.

Commit:
Fix Hive tests

We used to start multiple SparkSessions with Hive support.
Instead, now we explicitly pass in the SparkSession associated
with the TestHive singleton and update all the places that
try to instantiate HiveContext in tests. The error messages
here are particularly obscure.

SparkQA commented:
**[Test build #57232 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57232/consoleFull)** for PR 12746 at commit [`c4e48dc`](https://github.com/apache/spark/commit/c4e48dc2112011893d345824f2c88c97b7c5caef).

SparkQA commented:
**[Test build #57232 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57232/consoleFull)** for PR 12746 at commit [`c4e48dc`](https://github.com/apache/spark/commit/c4e48dc2112011893d345824f2c88c97b7c5caef).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57232/
Test FAILed.

Commit:
Fix style

SparkQA commented:
**[Test build #57233 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57233/consoleFull)** for PR 12746 at commit [`e61aa8a`](https://github.com/apache/spark/commit/e61aa8a10c5cca0615dea5544f0060781c385f89).

andrewor14 commented:
retest this please?? Why did you run all the unit tests??

SparkQA commented:
**[Test build #57234 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57234/consoleFull)** for PR 12746 at commit [`e61aa8a`](https://github.com/apache/spark/commit/e61aa8a10c5cca0615dea5544f0060781c385f89).

SparkQA commented:
**[Test build #57234 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57234/consoleFull)** for PR 12746 at commit [`e61aa8a`](https://github.com/apache/spark/commit/e61aa8a10c5cca0615dea5544f0060781c385f89).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57234/
Test PASSed.

SparkQA commented:
**[Test build #57233 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57233/consoleFull)** for PR 12746 at commit [`e61aa8a`](https://github.com/apache/spark/commit/e61aa8a10c5cca0615dea5544f0060781c385f89).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57233/
Test PASSed.

rxin commented:
Merging in master to unblock Andrew's follow-up pr. @davies can you do some posthoc review on this?

davies commented on python/pyspark/shell.py:
```diff
@@ -61,7 +62,7 @@
     platform.python_build()[0],
     platform.python_build()[1]))
-print("SparkContext available as sc, %s available as sqlContext." % sqlContext.__class__.__name__)
+print("SparkSession available as 'spark'.")
```
Should we mention that how to get a SparkContext from SparkSession?

davies commented:
LGTM

nchammas commented on python/pyspark/sql/session.py:
```diff
@@ -0,0 +1,525 @@
+
+
+def _monkey_patch_RDD(sparkSession):
+    def toDF(self, schema=None, sampleRatio=None):
```
Technically not your change @andrewor14 (I see you just moved this line), but it's `samplingRatio` everywhere else in the code. See: [samplingRatio](https://github.com/apache/spark/search?utf8=%E2%9C%93&q=samplingratio&type=Code) vs. [sampleRatio](https://github.com/apache/spark/search?utf8=%E2%9C%93&q=sampleratio&type=Code)

Should I open a PR to fix this?

rxin replied:
Please do.

rxin replied:
ah it would break APIs - maybe it's not worth it at this point.

nchammas replied:
Oh, it looks like a "non-public" method (which is implied in Python by the leading underscore in `_monkey_patch_RDD`) so I thought it would be OK to change.

Are you sure this counts as an API break? I did a search for `_monkey_patch_RDD` and, unless users are meant to use this directly, it looks like an internal-only method.

rxin replied:
monkey patch is not public, but the method being added via monkey patch becomes public (rdd.toDF)

nchammas replied:
Either way I guess this is not a big deal really.

nchammas replied:
Ahhhhh... OK I get it now.
